src/services/completions.ts

```diff
@@ -11,6 +11,22 @@ namespace ts.Completions {
     }
     export type Log = (message: string) => void;
 
+    /**
+     * Special values for `CompletionInfo['source']` used to disambiguate
+     * completion items with the same `name`. (Each completion item must
+     * have a unique name/source combination, because those two fields
+     * comprise `CompletionEntryIdentifier` in `getCompletionEntryDetails`.
+     *
+     * When the completion item is an auto-import suggestion, the source
+     * is the module specifier of the suggestion. To avoid collisions,
+     * the values here should not be a module specifier we would ever
+     * generate for an auto-import.
+     */
+    export enum CompletionSource {
+        /** Completions that require `this.` insertion text */
+        ThisProperty = "ThisProperty/"

```

Comment:
is "/" intentional at the end?

Reply:
Yeah, it’s explained in the comment—we need these values to be impossible values for an auto-import specifier. If someone had a folder named `ThisProperty` and a `baseUrl` set, we could conceivably generate an auto-import suggestion whose `source` is `ThisProperty`.